### PR TITLE
Integrate WYSIWYG editor

### PR DESCRIPTION
**Pull request by jmartinesp**
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

Adds the `io.element.android:wysiwyg` library, which features the new WYSIWYG editor. Modified `MessageComposerFragment` to be able to toggle between the current composer and the new one using a labs flag.

## Motivation and context

We want to integrate the initial version of the WYSIWIG editor into the real app to be able to send rich text messages without having to use markdown.

## Screenshots / GIFs

|Before|After|
|-|-|
|Screenshot_1664984239|Screenshot_1664984145|

## Tests

- Go to Settings > Labs, toggle 'Enable rich text editor'.
- Open any room or DM.
- Type any text using formatting and send it.
- You can also test if formatted text is sent for message edits, quotes, replies, etc.

## Tested devices

- [x] Physical
- [x] Emulator
- OS version(s): 11, 12

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [x] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [x] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/home/room/detail/composer/MessageComposerView.kt [hunk 1]
```diff
@@ -40,7 +46,7 @@ class MessageComposerView @JvmOverloads constructor(
         context: Context,
         attrs: AttributeSet? = null,
         defStyleAttr: Int = 0
-) : ConstraintLayout(context, attrs, defStyleAttr) {
+) : ConstraintLayout(context, attrs, defStyleAttr), MessageComposer {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/composer/MessageComposerView.kt:49]: For clarity, I would rename this class from `MessageComposerView` to `MessageComposerLayout` (or `MessageComposerLegacyLayout`) and your new interface could take the name `MessageComposerView`. WDYT?

File: vector/src/main/java/im/vector/app/features/home/room/detail/composer/MessageComposerViewModel.kt [hunk 2]
```diff
@@ -244,6 +251,21 @@ class MessageComposerViewModel @AssistedInject constructor(
...
+                            // Send the text message to the room, without markdown
+                            if (state.rootThreadEventId != null) {
+                                room.relationService().replyInThread(
+                                        rootThreadEventId = state.rootThreadEventId,
+                                        replyInThreadText = parsedCommand.message,
+                                        formattedText = parsedCommand.formattedMessage,
+                                        autoMarkdown = false
+                                )
+                            } else {
+                                room.sendService().sendFormattedTextMessage(parsedCommand.message.toString(), parsedCommand.formattedMessage)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/composer/MessageComposerViewModel.kt:264]: For parity with the `if` block, maybe use named param, and 1 per line.

File: vector/src/main/java/im/vector/app/features/home/room/detail/composer/RichTextComposerView.kt [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+import im.vector.app.databinding.ViewRichTextMenuButtonBinding
+import io.element.android.wysiwyg.InlineFormat
+
+class RichTextComposerView @JvmOverloads constructor(
+        context: Context,
+        attrs: AttributeSet? = null,
+        defStyleAttr: Int = 0
+) : ConstraintLayout(context, attrs, defStyleAttr), MessageComposer {
+
+    val views: ComposerRichTextLayoutBinding
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/composer/RichTextComposerView.kt:52]: Could be `private` I think.

**Review by bmarty — COMMENTED** [review]
Nice work!
Quick smoke test with the RTE enabled, seems to work fine.
The design will be updated? Could be nice to have something similar Element Web with popup:

image

**Comment by jmartinesp** [comment]
> Nice work! Quick smoke test with the RTE enabled, seems to work fine. The design will be updated? Could be nice to have something similar Element Web with popup:
> 
> image

The design that was chosen will keep this layout, we we thought about using popups but they'd clash with Android native ones (copy, paste, select all, etc.) and there might not be room for all the options to include in the future.

The design will be updated in a future PR, adding background and tint colors to the buttons and other sizes and appearance adjustments.

File: vector/src/main/java/im/vector/app/core/extensions/EditText.kt [hunk 4]
```diff
@@ -66,7 +67,7 @@ fun EditText.setTextIfDifferent(newText: CharSequence?): Boolean {
         // Previous text is the same. No op
         return false
     }
-    setText(newText)
+    setText(newText, TextView.BufferType.SPANNABLE)
```
[the remark this answers is not in the document]
Comment (jmartinesp) [reply vector/src/main/java/im/vector/app/core/extensions/EditText.kt]: Actually, this change wasn't needed, I forgot to revert a test. Great catch!

File: vector/src/main/res/layout/view_rich_text_menu_button.xml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<ImageButton xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="48dp"
+    android:layout_height="48dp"
+    android:layout_marginHorizontal="2dp"
+    android:background="@android:color/transparent"
+    android:contentDescription="@string/app_name">
+    <!-- The contentDescription attr is populated programmatically. This is just to fix lint issues. -->
```
[the remark this answers is not in the document]
Comment (jmartinesp) [reply vector/src/main/res/layout/view_rich_text_menu_button.xml:8]: As replied above, the final designs will be implemented on another PR if that's fine.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt [hunk 6]
```diff
@@ -124,19 +124,23 @@ internal class LocalEchoEventFactory @Inject constructor(
             roomId: String,
             targetEventId: String,
             newBodyText: CharSequence,
+            newBodyFormattedText: CharSequence?,
             newBodyAutoMarkdown: Boolean,
             msgType: String,
             compatibilityText: String
     ): Event {
+        val content = if (newBodyFormattedText != null) {
+            TextContent(newBodyText.toString(), newBodyFormattedText.toString()).toMessageTextContent(msgType)
```
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt:133]: Yes, you are right. Thanks

**Comment by andybalaam** [comment]
@bmarty since @jmartinesp is away, would you be able to merge this? Thanks!

**Pull request merged**